src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs

```diff
@@ -0,0 +1,258 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using System.Xml.Linq;
+using Hazelcast.Core;
+using Hazelcast.Networking;
+using Hazelcast.Testing;
+using Hazelcast.Testing.Logging;
+using Microsoft.Extensions.Logging;
+using NUnit.Framework;
+
+namespace Hazelcast.Tests.Networking
+{
+    public class RegressionWithRealNetworkTest : MultiMembersRemoteTestBase
+    {
+        private IDisposable HConsoleForTest()
+            => HConsole.Capture(options => options
+                .ClearAll()
+                .Configure<HConsoleLoggerProvider>().SetPrefix("LOG").SetMaxLevel()
+                .Configure().SetMinLevel().EnableTimeStamp(origin: DateTime.Now)
+                .Configure(this).SetMaxLevel().SetPrefix("TEST")
+            );
+
+        [TearDown]
+        public async Task TearDown()
+        {
+            //since tests are network related may need fresh members on cases.
+            await MembersOneTimeTearDown();
+        }
+
+        // this test validates that a client that starts connecting to a cluster that is not yet ready,
+        // will eventually succeeds and connect once the cluster becomes ready.
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task TestClientConnectionBeforeServerReady(bool smartRouting)
+        {
+            using var _ = HConsoleForTest();
+
+            var clientTask = CreateAndStartClientAsync(options =>
+            {
+                options.Networking.SmartRouting = smartRouting;
+                options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = int.MaxValue;
+                options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
```

Comment:
`Async` and `Sync` modes are same at the .Net client.

Reply:
I mean why don't you also have the one with `ReconnectMode.ReconnectSync`

Reply:
Because they behave same but I added anyway, incase of we can change their implementation in the future. Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/614/commits/93423d885cc9f88fba8896ebe4b1990e801f3a45